Pinta.Effects/Effects/DitheringEffect.cs

```diff
@@ -0,0 +1,149 @@
+using System;
+using System.Collections.Immutable;
+using Cairo;
+using Pinta.Core;
+using Pinta.Gui.Widgets;
+
+namespace Pinta.Effects;
+
+public sealed class DitheringEffect : BaseEffect
+{
+	public override string Name => Translations.GetString ("Dithering");
+	public override bool IsConfigurable => true;
+	// TODO: Icon
+	public override string EffectMenuCategory => Translations.GetString ("Color");
+	public DitheringData Data => (DitheringData) EffectData!; // NRT - Set in constructor
+
+	public override bool IsTileable => false;
+
+	public DitheringEffect ()
+	{
+		EffectData = new DitheringData ();
+	}
+
+	public override void LaunchConfiguration ()
+	{
+		EffectHelper.LaunchSimpleEffectDialog (this);
+	}
+
+	private sealed record DitheringSettings (
+		ErrorDiffusionMatrix diffusionMatrix,
+		ImmutableArray<ColorBgra> palette,
+		int sourceWidth,
+		int sourceHeight);
+
+	private DitheringSettings CreateSettings (ImageSurface src)
+		=> new (
+			diffusionMatrix: ErrorDiffusionMatrix.GetPredefined (Data.DiffusionMatrix),
+			palette: PaletteHelper.GetPredefined (Data.PaletteChoice),
+			sourceWidth: src.Width,
+			sourceHeight: src.Height
+		);
+
+	protected override void Render (ImageSurface src, ImageSurface dest, RectangleI roi)
+	{
+		DitheringSettings settings = CreateSettings (src);
+
+		ReadOnlySpan<ColorBgra> src_data = src.GetReadOnlyPixelData ();
+		Span<ColorBgra> dst_data = dest.GetPixelData ();
+
+		for (int y = roi.Top; y <= roi.Bottom; y++) {
+			for (int x = roi.Left; x <= roi.Right; x++) {
+				int currentIndex = y * settings.sourceWidth + x;
+				dst_data[currentIndex] = src_data[currentIndex];
+			}
+		}
+
+		for (int y = roi.Top; y <= roi.Bottom; y++) {
+
+			for (int x = roi.Left; x <= roi.Right; x++) {
+
+				int currentIndex = y * settings.sourceWidth + x;
+				ColorBgra originalPixel = dst_data[currentIndex];
+				ColorBgra closestColor = FindClosestPaletteColor (settings.palette, originalPixel);
+
+				dst_data[currentIndex] = closestColor;
+
+				int errorRed = originalPixel.R - closestColor.R;
+				int errorGreen = originalPixel.G - closestColor.G;
+				int errorBlue = originalPixel.B - closestColor.B;
+
+				for (int r = 0; r < settings.diffusionMatrix.Rows; r++) {
+
+					for (int c = 0; c < settings.diffusionMatrix.Columns; c++) {
+
+						var weight = settings.diffusionMatrix[r, c];
+
+						if (weight <= 0)
+							continue;
+
+						PointI thisItem = new (
+							X: x + c - settings.diffusionMatrix.ColumnsToLeft,
+							Y: y + r
+						);
+
+						if (thisItem.X < roi.Left || thisItem.X >= roi.Right)
+							continue;
+
+						if (thisItem.Y < roi.Top || thisItem.Y >= roi.Bottom)
+							continue;
+
+						if (thisItem.X < 0 || thisItem.X >= settings.sourceWidth)
```

Comment:
This second set of checks should be unnecessary since the rectangle should be inside the image bounds

Reply:
At first I added it because I wasn't sure if the ROIs were guaranteed to be inside the image bounds, but in the light of this I removed it.